src/lib.rs

```diff
@@ -182,6 +182,49 @@ impl iter::FromIterator<char> for SmolStr {
     }
 }
 
+fn build_from_str_iter<T>(mut iter: impl Iterator<Item=T>) -> SmolStr
+where
+    T: AsRef<str>,
+    std::string::String: std::iter::Extend<T>,
+{
+    use std::io::prelude::*;
+
+    let mut len = 0;
+    let mut buf = [0u8; INLINE_CAP];
+    while let Some(slice) = iter.next() {
+        let slice = slice.as_ref();
+        let size = slice.len();
+        if size + len > INLINE_CAP {
+            let mut heap = String::with_capacity(size + len);
+            heap.push_str(std::str::from_utf8(&buf[..len]).unwrap());
+            heap.push_str(&slice);
+            heap.extend(iter);
+            return SmolStr(Repr::Heap(heap.into_boxed_str().into()));
+        }
+        (&mut buf[len..]).write_all(slice.as_bytes()).unwrap();
```

Comment:
Isn't this `.copy_from_slice`?

Reply:
Nah, that panics when the sizes are different.

Reply:
Since you've already checked that `size + len <= INLINE_CAP`, can't you do `buf[len..len+size]` (alternatively `buf[len..][..size]` `.copy_from_slice(...)`?

Reply:
I changed it to the more explicit version, mostly because it's easier to understand :)